publish-externals/src/index.ts

```diff
@@ -10,28 +12,89 @@ const ExternalPublications = [
-          await publication.handler(frontMatter, content)
-          console.log(`[${publication.name}] Success ${frontMatter.slug} ... succeeded ✅`)
-        } catch (error) {
-          errors.push(error)
-          console.error(`[${publication.name}] Failure ${frontMatter.slug} ... failed 💀`)
+      await publish(file.replace("blog/", ""))
+    }
+  }
+
+  // publish blogs/updates which were not published before
+  for (const blog of blogs) {
+    if (!blog.lastUpdatePublished || !blog.published) {
+      await publish(blog.blogName)
+    }
+  }
+}
+
+const publish = async (file: string) => {
+  const blogs = snapshot.snapshot || []
+  const errors = []
+  const filePath = path.join(__dirname, "../../", `blog/${file}`)
+  const {frontMatter, content} = extractFrontMatterAndContent(filePath)
+  for (let publication of ExternalPublications) {
+    console.log(`[${publication.name}] ${frontMatter.slug} ... publishing ⏳`)
+    try {
+      await publication.handler(frontMatter, content)
+      // Update fields and if blog does not exist in snapshot, add it
+      const inSnapshot = blogs.findIndex((blog) => blog.blogName === file)
+      if (inSnapshot !== -1) {
+        // ensure only to publish to platforms mentioned in snapshot
+        if (blogs[inSnapshot].platforms.includes(publication.name)) {
+          blogs[inSnapshot].published = true
+          blogs[inSnapshot].lastUpdatePublished = true
+          blogs[inSnapshot].lastSuccessfulPublishedAt = new Date().toUTCString()
+          await writeSnapshot(blogs)
         }
+      } else {
+        blogs.push({
+          blogName: file,
+          published: true,
+          platforms: ["Hashnode", "Dev.to"],
+          lastUpdatePublished: true,
+          lastSuccessfulPublishedAt: new Date().toUTCString(),
+        })
+        await writeSnapshot(blogs)
       }
+      console.log(`[${publication.name}] Success ${frontMatter.slug} ... succeeded ✅`)
+    } catch (error) {
+      errors.push(error)
+      console.error(`[${publication.name}] Failure ${frontMatter.slug} ... failed 💀`)
     }
   }
-
   if (errors.length !== 0) {
+    // Update fields and if blog does not exist in snapshot, add it
+    const inSnapshot = blogs.findIndex((blog) => blog.blogName === file)
+    if (inSnapshot !== -1) {
+      blogs[inSnapshot].lastUpdatePublished = false
+      await writeSnapshot(blogs)
+    } else {
+      blogs.push({
+        blogName: file,
+        platforms: ["Hashnode", "Dev.to"],
+        published: false,
+        lastUpdatePublished: false,
+      })
+      await writeSnapshot(blogs)
+    }
+
     console.error(errors)
     throw new Error("Publishing failed because of one or more errors")
   }
 }
 
+const writeSnapshot = async (snapshot: any) => {
+  try {
+    fs.writeFileSync(path.join(__dirname, "../", "snapshot.json"), JSON.stringify({snapshot}, null, 2), (err) => {
+      if (err) {
+        console.error(err)
+      }
+    })
+  } catch (err) {
+    console.error(err)
```

Comment:
Incase of failure we should print the snapshot so that we can manually handle it if required.